src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -400,6 +402,28 @@ public virtual async Task<ConnectionCompleteParams> Connect(ConnectParams connec
             return completeParams;
         }
 
+        private async Task<ConnectionCompleteParams> TryOpenConnectionWithRetry(ConnectionInfo connectionInfo, ConnectParams connectionParams)
+        {
+            Boolean isRetriable = true;
+            int counter = 0;
+            ConnectionCompleteParams response = null;
+            while (isRetriable && counter <= MaxTries)
+            {
+                response = await TryOpenConnection(connectionInfo, connectionParams);
+                // If a serverless database is sleeping, it will return this error number and will need to be retried.
+                if (response?.ErrorNumber == 40613)

```

Comment:
@shueybubbles @Matteo-T Is this what we do in SSMS as well? Or is there some built in functionality in SqlClient we can use for this?

Reply:
I searched around in both SSMS and SMO but didn't find any reference to 40613, serverless or sleeping...

Reply:
It should be done by SSMS as well.. since serverless DBs are new, it's always better to let them resume and retry silently..!

Reply:
Also to add, the overall connection time may exceed timeout limit of 30 seconds, and if we want to stay within timeout limit, this change should be avoided. But if we can prompt user that server is resuming timely, and continue with connection, that should be acceptable I think.

So my ask would be is there a way to provide user feedback to end user that "The server is resuming, please wait while we continue to connect .. "?

Reply:
I would need to design a way to relay information from STS back to ADS, which will require some work to do.

Reply:
Chatted with @shueybubbles offline, it looks like SMO/SSMS isn't doing any specific retry logic for Azure DB serverless (there's one for synapse though)

@smartguest Could you try connecting to a sleeping serverless instance with SSMS to verify what the behavior is there? It'll be good to verify what the behavior is across our tools for comparison.